[master] Add tests for minion execution module

techhat opened the pull request:
### What does this PR do?
Add tests for minion execution module

### What issues does this PR fix or reference?
References #64697

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
Yes

s0undt3ch commented on tests/pytests/integration/minion/test_minion_module.py:
```diff
@@ -0,0 +1,93 @@
+from saltfactories.utils import random_string
+
+
+@pytest.fixture(scope="module")
+def module_master(salt_master_factory):
+    yield salt_master_factory
+
+
+@pytest.fixture(scope="module")
+def master_minion(module_master):
+    config_overrides = {
+        "root_dir": f"/tmp/stsuite/{module_master.id}",
+        "return_retry_timer_max": 0,
+        "return_retry_timer": 5,
+        "return_retry_tries": 30,
+    }
+    factory = module_master.salt_minion_daemon(
+        module_master.id,
+        overrides=config_overrides,
+        extra_cli_arguments_after_first_start_failure=["--log-level=info"],
+    )
+    factory.after_terminate(
+        pytest.helpers.remove_stale_minion_key, module_master, factory.id
+    )
+
+    with factory.started():
+        yield factory
```
```suggestion
@pytest.fixture(scope="module")
def master_minion(salt_master):
    config_overrides = {
        "root_dir": salt_master.config["root_dir"],
        "return_retry_timer_max": 0,
        "return_retry_timer": 5,
        "return_retry_tries": 30,
    }
    factory = salt_master.salt_minion_daemon(
        salt_master.id,
        overrides=config_overrides,
        extra_cli_arguments_after_first_start_failure=["--log-level=info"],
    )
    factory.after_terminate(
        pytest.helpers.remove_stale_minion_key, salt_master, factory.id
    )

    with factory.started():
        yield factory
```

s0undt3ch commented on tests/pytests/integration/minion/test_minion_module.py:
```diff
@@ -0,0 +1,93 @@
+        yield factory
+
+
+@pytest.fixture(scope="module")
+def second_minion(module_master, salt_minion_id):
+    config_overrides = {
+        "return_retry_timer_max": 0,
+        "return_retry_timer": 5,
+        "return_retry_tries": 30,
+    }
+    factory = module_master.salt_minion_daemon(
+        salt_minion_id,
+        extra_cli_arguments_after_first_start_failure=["--log-level=info"],
+    )
+    factory.after_terminate(
+        pytest.helpers.remove_stale_minion_key, module_master, factory.id
+    )
+
+    with factory.started():
+        yield factory
```
```suggestion
@pytest.fixture(scope="module")
def second_minion(salt_master, salt_minion_id):
    config_overrides = {
        "return_retry_timer_max": 0,
        "return_retry_timer": 5,
        "return_retry_tries": 30,
    }
    factory = salt_master.salt_minion_daemon(
        salt_minion_id,
        extra_cli_arguments_after_first_start_failure=["--log-level=info"],
    )
    factory.after_terminate(
        pytest.helpers.remove_stale_minion_key, salt_master, factory.id
    )

    with factory.started():
        yield factory
```

s0undt3ch commented on tests/pytests/integration/minion/test_minion_module.py:
```diff
@@ -0,0 +1,93 @@
+    salt_run_cli,
+    salt_key_cli,
+):
+    master_dir = Path(f"/tmp/stsuite/{salt_master.id}")
+    shutil.copy(
+        master_dir / "pki" / "minion.pub",
+        master_dir / "pki" / "minions" / f"{salt_master.id}",
+    )
+    master_minion.start()
```
```suggestion
    master_pki_dir = salt_master.config["pki_dir"]
    shutil.copy(
        master_pki_dir / "minion.pub",
        master_pki_dir / "minions" / f"{salt_master.id}",
    )
    master_minion.start()
```

s0undt3ch commented on tests/pytests/integration/minion/test_minion_module.py:
```diff
@@ -0,0 +1,86 @@
+        "return_retry_tries": 30,
+    }
+    factory = salt_master.salt_minion_daemon(
+        salt_minion_id,
```
```suggestion
        salt_minion_id,
```
During integration tests, there's a a salt master and a salt minion running, `salt_minion_id` is whats used for the minion ID.

Why are we reusing that `salt_minion_id` as the second master here?

techhat replied:
If you look in the `master_minion` function you'll see that I'm using the master ID as the ID for the minion that's started up with it, which allows them to share the same directory in the test suite. That means the generated `salt_minion_id` is free to use for the second minion.

s0undt3ch replied:
That is only true if no other tests run.
If other tests run, for example, before this one, the minion with the ID of the value of `salt_minion_id` would already be running, and reusing that ID would mean two minions with the same ID(Don't really know how salt handles that)

s0undt3ch commented on tests/pytests/integration/minion/test_minion_module.py:
```diff
@@ -0,0 +1,86 @@
+
+    salt_cli.run("minion.restart", minion_tgt=second_minion.id, timeout=1)
+
+    start_pattern = "salt/minion/flay/start"
```
```suggestion
    start_pattern = "salt/minion/flay/start"
```
There's a minion ID `flay` defined somewhere?

s0undt3ch commented on tests/pytests/integration/minion/test_minion_module.py:
```diff
@@ -0,0 +1,86 @@
+
+
+@pytest.fixture(scope="module")
+def second_minion(salt_master, salt_minion_id):
```
```suggestion
def second_minion(salt_master):
```

s0undt3ch commented on tests/pytests/integration/minion/test_minion_module.py:
```diff
@@ -0,0 +1,86 @@
+        "return_retry_tries": 30,
+    }
+    factory = salt_master.salt_minion_daemon(
+        salt_minion_id,
```
```suggestion
        random_string("2nd-minion-"),
```
```python
from saltfactories.utils import random_string
```

s0undt3ch commented on tests/pytests/integration/minion/test_minion_module.py:
```diff
@@ -0,0 +1,88 @@
+
+
+@pytest.fixture(scope="module")
+def second_minion(salt_master, salt_minion_id):
```
```suggestion
def second_minion(salt_master):
```

dwoz commented:
@techhat Ping

dwoz committed:
Fix MinionManager skipping connect_master when file_client is local

MinionManager._connect_minion had an inverted boolean guard: the original
PR's condition skipped connect_master when use_master_when_local was True,
and called it when use_master_when_local was False — the opposite of the
intended behaviour. Align with the pattern in resolve_dns() and
Minion.__init__: skip connect_master only when file_client is local AND
use_master_when_local is False.

Add regression tests for both the fixed case (local + use_master=False
must not connect) and the expected-connect case (local + use_master=True
must connect).

Fixes #57866
Fixes #64952

twangboy commented:
This should go against the earliest branch where the bug exists, likely 3006.x
